tensorboard/plugins/graph/graph_util.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto import function_pb2
+
+
+class _ProtoListNonUniqueKeyError(Exception):
+  pass
+
+
+class _SameKeyDiffContentError(Exception):
+  pass
+
+
+def _safe_copy_proto_list_values(dst_proto_list, src_proto_list, get_key):
+  """Safely copies the value from src_proto_list to dst_proto_list.
+
+  First checks whether keys in both dst_proto_list and src_proto_list are
```

Comment:
As written, this paragraph defines _how_ the function is implemented,
not _what_ its contract is. The person reading these docs doesn’t care
what happens first or second (or they’d look at the code); they need to
know what the function _requires_ about its arguments and what it
_ensures_ about its result.

Suggest something like:

> ```python
>   """Safely merge values from `src_proto_list` into `dst_proto_list`.
>
>   Each element in `dst_proto_list` must be mapped by `get_key` to a key
>   value that is unique within that list; likewise for `src_proto_list`.
>   If an element of `src_proto_list` has the same key as an existing
>   element in `dst_proto_list`, then the elements must also be equal.
>
>   ...
>   """
> ```

The precise exceptions raised don’t need to be specified in the first
paragraph; they’re already in the “Raises” section.